sensor/tests/resource/service/service_test.go

```diff
@@ -0,0 +1,352 @@
+}
+
+var _ suite.SetupAllSuite = &DeploymentExposureSuite{}
+var _ suite.TearDownTestSuite = &DeploymentExposureSuite{}
+
+func (s *DeploymentExposureSuite) TearDownTest() {
+	// Clear any messages received in fake central during the test run
+	s.testContext.GetFakeCentral().ClearReceivedBuffer()
+}
+
+func (s *DeploymentExposureSuite) SetupSuite() {
+	if testContext, err := resource.NewContext(s.T()); err != nil {
+		s.Fail("failed to setup test context: %s", err)
+	} else {
+		s.testContext = testContext
+	}
+}
+
+func (s *DeploymentExposureSuite) Test_ClusterIpPermutation() {
+	s.testContext.RunWithResourcesPermutation(
+		[]resource.YamlTestFile{
+			NginxDeployment,
+			NginxServiceClusterIP,
+		}, "Cluster IP", func(t *testing.T, testC *resource.TestContext, _ map[string]k8s.Object) {
+			// Test context already takes care of creating and destroying resources
+			time.Sleep(2 * time.Second)
+			assertLastDeploymentHasPortExposure(
+				t,
+				testC.GetFakeCentral().GetAllMessages(),
+				[]*storage.PortConfig{
+					{
+						Protocol:      "TCP",
+						ContainerPort: 9376,
+						Exposure:      storage.PortConfig_INTERNAL,
+						ExposureInfos: []*storage.PortConfig_ExposureInfo{
+							{
+								ServiceName: "nginx-svc-cluster-ip",
+								ServicePort: 80,
+								Level:       storage.PortConfig_INTERNAL,
+							},
+						},
+					},
+				},
+				[]*storage.Alert{},
+			)
+			testC.GetFakeCentral().ClearReceivedBuffer()
+		},
+	)
+}
+
+func (s *DeploymentExposureSuite) Test_NodePortPermutation() {
+	s.testContext.RunWithResourcesPermutation(
+		[]resource.YamlTestFile{
+			NginxDeployment,
+			NginxServiceNodePort,
+		}, "NodePort", func(t *testing.T, testC *resource.TestContext, _ map[string]k8s.Object) {
+			// Test context already takes care of creating and destroying resources
+			time.Sleep(2 * time.Second)
+			assertLastDeploymentHasPortExposure(
+				t,
+				testC.GetFakeCentral().GetAllMessages(),
+				[]*storage.PortConfig{
+					{
+						Protocol:      "TCP",
+						ContainerPort: 80,
+						Exposure:      storage.PortConfig_NODE,
+						ExposureInfos: []*storage.PortConfig_ExposureInfo{
+							{
+								ServiceName: "nginx-svc-node-port",
+								ServicePort: 80,
+								NodePort:    30007,
+								Level:       storage.PortConfig_NODE,
+							},
+						},
+					},
+				},
+				[]*storage.Alert{
+					{
+						Policy: &storage.Policy{
+							Name: "test-service",
+						},
+						State: storage.ViolationState_ACTIVE,
+					},
+				},
+			)
+			testC.GetFakeCentral().ClearReceivedBuffer()
+		},
+	)
+}
+
+func (s *DeploymentExposureSuite) Test_LoadBalancerPermutation() {
+	s.testContext.RunWithResourcesPermutation(
+		[]resource.YamlTestFile{
+			NginxDeployment,
+			NginxServiceLoadBalancer,
+		}, "LoadBalancer", func(t *testing.T, testC *resource.TestContext, _ map[string]k8s.Object) {
+			// Test context already takes care of creating and destroying resources
+			time.Sleep(2 * time.Second)
```

Comment:
Potential to be a source of flakiness . Is this necessary? Is there any non-flaky alternative for this?

Reply:
Since re-sync is time-based (and set to 1s in these tests), the easiest way to make sure that re-sync already kicked-in is simply to wait the re-sync time and check the last message received. All tests do that at the moment. 

However, as I write this, I realized that there might be a more clever way to check for incoming messages with certain values with a timeout. Since it's a bigger topic in the test context library, I wouldn't expect it to be done in this PR. But I can follow up with that and propose PR for it.

Reply:
Just out of curiosity, how are you planning to wait for the re-sync update events? AFAIK there isn't any indicator to differentiate re-sync events from legitimate update events.